src/lib/SPI/Repository/Value/Filter/FilteringCriterion.php

```diff
@@ -0,0 +1,16 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace Ibexa\Platform\SPI\Repository\Value\Filter;
+
+/**
+ * Marker for Content & Location filtering Criterion.
```

Comment:
Does it mean a Criterion that supports both should only implement that interface ? What is its actual meaning ?

Reply:
> Does it mean a Criterion that supports both should only implement that interface ? What is its actual meaning ?

That's the initial idea - if it supports both, it should implement that interface. 
But I need to apply this to all Criteria to see if keeping this common interface makes sense.

Reply:
@alongosz A bit late to the party, but I did start at creating own separate Persistence Criteria some time back, for the purpose of usage in Trash filtering as well as for the methods proposed here.

Besides separation, another reason for doing so was to experiment with rather using joins instead of subqueries to make sure we don't hit performance issues. Even if newer mysql/mariadb versions are supposed to optimize this they seem to have plenty of cases where they won't be able to.

https://github.com/ezsystems/ezpublish-kernel/commit/90c954f52bc6564508db642522a137c4bfe96abf  _(see Query/Readme.md at the end of diff for overview)_

Reply:
> A bit late to the party, but FYI started at creating own separate Persistence Criteria some time back, for the purpose of usage in Trash filtering
> (...)
> [ezsystems/ezpublish-kernel@90c954f](https://github.com/ezsystems/ezpublish-kernel/commit/90c954f52bc6564508db642522a137c4bfe96abf)

AFAICS the commit is about Criterion handlers, not the Criteria themselves, right? If you planned to add separate Criteria also, then this wouldn't be the best direction IMHO. Criterion itself should not be coupled with any storage. It can be coupled with persistence abstraction for sure, but I have a feeling our persistence abstraction is too coupled with Legacy Storage TBH ;)

> Besides separation, another reason for doing so was to experiment with rather using joins instead of subqueries to make sure we don't hit performance issues. Even if newer mysql/mariadb versions are supposed to optimize subqueries, they seem to have plenty of cases where they won't be able to.

Criterion Handling is going to be separate here for Content/Location filtering, so it can utilize joins when needed, but only when underlying schema makes it correct (e.g.: I'd like to avoid programmatic cleanup of redundant data).

Reply:
> AFAICS the commit is about Criterion handlers, not the Criteria themselves, right? If you planned to add separate Criteria also

yes, indeed own set of Criteria would be one of several followup there. But not due to coupling as you jump to conclusion on, rather:
- proper type hints on the new api's on what is allowed to use and not, like Adam already stated to do on URLService for the same reasons.

Beyond that rethink how CriteriaHandlers are done to see if we could make them use join s instead.

Anyway, good luck.